src/Codec/Archive/ZTar/Zip.hs

```diff
@@ -57,21 +58,26 @@ create :: FilePath -- ^ archive to create
        -> IO ()
 create archive base paths = do
   archive' <- makeAbsolute archive
-  withCurrentDirectory base $ Zip.createArchive archive' $ mapM_ insert paths
+  withCurrentDirectory base $ do
+    files <- concatMapM search paths
+    Zip.createArchive archive' $ mapM_ insertFile files
   where
-    insert path = do
-      isFile <- liftIO $ doesFileExist path
-      isDir <- liftIO $ doesDirectoryExist path
+    search :: FilePath -> IO [FilePath]
+    search path = do
+      isFile <- doesFileExist path
+      isDir <- doesDirectoryExist path
       if
-        | isFile -> insertFile path
-        | isDir -> insertDir path
+        | isFile -> pure [path]
+        | isDir -> searchDir path
         | otherwise -> fail $ "Path does not exist: " ++ path
+    searchDir :: FilePath -> IO [FilePath]
+    searchDir path =
+      let mkPath = if path == "." then id else (path </>)
+      in concatMapM (search . mkPath) =<< listDirectory path
     insertFile path = do
       path' <- Zip.mkEntrySelector path
       Zip.loadEntry Zip.BZip2 path' path
-    insertDir path =
-      let mkPath = if path == "." then id else (path </>)
-      in mapM_ (insert . mkPath) =<< liftIO (listDirectory path)
+    concatMapM f = liftM concat . mapM f
```

Comment:
why not just fmap?

Reply:
This was in the GHC proposal for adding concatMapM to base. Tell me what you want written and I'll change it; I'm not going to make guesses about the aesthetics.

Reply:
oh ok, that's fine. just curious

Reply:
I guess the reason is that `liftM` is not necessarily identical to `fmap` for all implementations of a Monad.

Reply:
makes sense. but isn't it a monad law that
```hs
m >>= return . f
```
is equivalent to
```hs
f <$> m
```
? i don't see why `liftM` shouldn't be considered equivalent to `fmap`

Reply:
You can make a Monad that doesn't obey the law, so for a `base` package, this implementation would make sense. This is in a specific monad anyway, so it doesn't need to generalize or account for the possibilities that monads don't obey these laws. You're welcome to change it; I just picked the first implementation of concatMapM I came across. This really doesn't matter to me.

The proposal is _quite_ old, so maybe there is some historical reason they wrote it like this.
https://gitlab.haskell.org/ghc/ghc/issues/2042